std-bits/base/src/main/java/org/enso/base/parser/NumberWithSeparators.java

```diff
@@ -0,0 +1,393 @@
+            : (decimal.isEmpty() ? Constants.NONE : decimal.charAt(0));
+
+    Optional<NumberWithSeparators> matched =
+        switch (thousands) {
+          case Constants.NONE -> matchForNone(decimals);
+          case Constants.UNKNOWN -> matchForUnknown(decimals);
+          case ',' -> Optional.of(COMMA_DOT);
+          case '.' -> Optional.of(DOT_COMMA);
+          case ' ' -> matchForSpace(decimals);
+          case '\'' -> matchForSwiss(decimals);
+          default -> Optional.empty();
+        };
+
+    if (matched.isEmpty()) {
+      throw new IllegalArgumentException("Invalid separators.");
+    }
+    return matched.get();
+  }
+
+  private static Optional<NumberWithSeparators> matchForNone(char decimal) {
+    return switch (decimal) {
+      case Constants.UNKNOWN -> Optional.of(NO_UNKNOWN);
+      case '.' -> Optional.of(NO_DOT);
+      case ',' -> Optional.of(NO_COMMA);
+      default -> Optional.empty();
+    };
+  }
+
+  private static Optional<NumberWithSeparators> matchForUnknown(char decimal) {
+    return switch (decimal) {
+      case Constants.UNKNOWN -> Optional.of(UNKNOWN);
+      case '.' -> Optional.of(UNKNOWN_DOT);
+      case ',' -> Optional.of(UNKNOWN_COMMA);
+      default -> Optional.empty();
+    };
+  }
+
+  private static Optional<NumberWithSeparators> matchForSpace(char decimal) {
+    return switch (decimal) {
+      case Constants.UNKNOWN -> Optional.of(SPACE_UNKNOWN);
+      case '.' -> Optional.of(SPACE_DOT);
+      case ',' -> Optional.of(SPACE_COMMA);
+      default -> Optional.empty();
+    };
+  }
+
+  private static Optional<NumberWithSeparators> matchForSwiss(char decimal) {
+    return switch (decimal) {
+      case Constants.UNKNOWN -> Optional.of(SWISS_UNKNOWN);
+      case '.' -> Optional.of(SWISS_DOT);
+      case ',' -> Optional.of(SWISS_COMMA);
+      default -> Optional.empty();
+    };
+  }
+
+  static class Constants {
+    static final char NONE = '\0';
+    static final char UNKNOWN = '\uFFFD';
+  }
+
+  static boolean isDigit(char c) {
+    return (c >= '0' && c <= '9');
+  }
+
+  private final char thousands;
+  private final char decimal;
+
+  NumberWithSeparators(char thousands, char decimal) {
+    this.thousands = thousands;
+    this.decimal = decimal;
+  }
+
+  public char getThousands() {
+    return thousands;
+  }
+
+  public char getDecimal() {
+    return decimal;
+  }
+
+  NumberParseResult parse(CharSequence value, int idx, boolean integer) {
+    var separators = Separators.parse(value, idx, integer);
+    if (separators == null) {
+      return new NumberParseFailure("Invalid separators.");
+    }
+
+    if (thousands != Constants.UNKNOWN && (integer || decimal != Constants.UNKNOWN)) {
+      // If we have a fixed format then we can parse the number.
+      return integer
+          ? parseFixedInteger(value, idx, separators.endIdx(), separators.first())
+          : parseFixed(value, idx, separators.endIdx(), separators.first(), separators.second());
+    }
```

Comment:
How is the value stored in `thousands` related to `separators.first()` here? Shouldn't we somehow check that they are the same?

Reply:
Oh, it's done in the function... Sorry